dash/_validate.py

```diff
@@ -0,0 +1,571 @@
+import collections
+import re
+
+from .development.base_component import Component
+from .dependencies import Input, Output, State, MATCH, ALLSMALLER
+from . import exceptions
+from ._utils import patch_collections_abc, _strings, stringify_id
+
+
+def validate_callback(app, layout, output, inputs, state):
+    is_multi = isinstance(output, (list, tuple))
+    validate_ids = not app.config.suppress_callback_exceptions
+
+    if layout is None and validate_ids:
+        # Without a layout, we can't do validation on the IDs and
+        # properties of the elements in the callback.
+        raise exceptions.LayoutIsNotDefined(
+            """
+            Attempting to assign a callback to the application but
+            the `layout` property has not been assigned.
+            Assign the `layout` property before assigning callbacks.
+            Alternatively, suppress this warning by setting
+            `suppress_callback_exceptions=True`
+            """
+        )
```

Comment:
Not in scope but I've run across the issue a few times where I've defined my callbacks before assigning layout. I wonder if it would be possible to relax that requirement. I think this is because we validate the callback ids on the spot, but if we were to accumulate and validate them on `run_server` this wouldn't be necessary.